pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go

```diff
@@ -194,12 +199,6 @@ func (dswp *desiredStateOfWorldPopulator) findAndAddNewPods() {
 	}
 
 	for _, pod := range dswp.podManager.GetPods() {
-		// Keep consistency of adding pod during reconstruction
-		if dswp.hasAddedPods && dswp.podStateProvider.ShouldPodContainersBeTerminating(pod.UID) {
```

Comment:
I don't know this subsystem in detail, but I doubt completely removing this is correct

The overall goal is to *keep* refreshing volumes *which have already been set up* as long as the containers they are mounted into are not yet terminated. I don't see those aspects considered here.

The thing we want to guard against is unnecessarily trying to set up *new* volumes for containers which have not yet started if the pod has gone into terminating state so those new containers will never be started. I don't see those aspects considered here either.

I think detailed guidance from sig-storage maintainers is needed to know how to accomplish both of the above goals simultaneously, and it's possible (likely, even) that the current volume setup / sync approach will need tweaking to allow that.

Reply:
Thanks for your feedback, @liggitt !

> The overall goal is to keep refreshing volumes which have already been set up as long as the containers they are mounted into are not yet terminated. I don't see those aspects considered here.

In this PR, I'm accounting for that by having `podPreviouslyProcessed()` start returning `false` if its been more than 60 seconds since `markPodProcessed()` was called. And this timeout works even if the Pod is in a terminating state and the loop that manages the Pod is stuck waiting for the Pod to terminate. Or, am I misunderstanding your comment?

> The thing we want to guard against is unnecessarily trying to set up new volumes for containers which have not yet started if the pod has gone into terminating state so those new containers will never be started. I don't see those aspects considered here either.

Ah, cool. I didn't take that into consideration. I will look into that.

Reply:
@liggitt I've pushed a new change that I hope addresses you feedback (assuming that I understand it correctly - apologies if I do not). I have also updated the PR description with a better summary of the change I'm proposing here. Thanks!

Reply:
@liggitt do you have any thoughts on my responses to your comments? Do you have any thoughts on how I could request guidance from the sig-storage folks? Thanks!

Reply:
> I think detailed guidance from sig-storage maintainers is needed

I don't the expertise to review this, we *really* need storage subsystem experts to review this change closely

Reply:
> In this PR, I'm accounting for that by having `podPreviouslyProcessed()` start returning `false` if its been more than 60 seconds since `markPodProcessed()` was called.

What is significant about 60 seconds? It is not clear to me how that scopes this change to exactly the containers and volumes we care about continuing to refresh

Reply:
Unless I'm misunderstanding, there are two loops currently being used:

1. The first, is that volume_manager.go creates a DesiredStateOfWorldPopulator [here](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/volumemanager/volume_manager.go#L213). It starts it [here](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/volumemanager/volume_manager.go#L294). It runs every [100 milliseconds](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/volumemanager/volume_manager.go#L64). Its not really clear to me why its 100 milliseconds. Anyway, as best as I can tell, this loop mostly runs and does nothing until its triggered by the second loop.

2. The second loop is what I'm calling the SyncPod loop. Its defined [here](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/kubelet.go#L2323). Once [every second](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/kubelet.go#L2328), a timer fires, and through a chain of calls, that causes markPodProcessingFailed() to be called.

Anyway, the first loop does nothing most times it run. However, after markPodProcessingFailed() is called, that causes the first loop to jump into action and update the projected volumes, at least as I understand it.

So, why did I pick 60 seconds? I picked it because it seemed that a 60 second timeout was quite a bit longer than the existing loops (100ms and 1s) and so re-processing projected volumes every 60 seconds wouldn't be a performance issue when we were already handling them once a second. However, 60 seconds also seems short enough that any projected volumes that need to be updated will not have gotten stale in less than 60 seconds.

So, yes, its a bit of a magic number. But the other loops seem like they have magic numbers as well, unless there is something I'm missing - and if thats the case, please let me know.

Reply:
@liggitt as I understand it you would be more willing to keep a refresh every 1s to match the syncPod periodicity during termination?

Reply:
I don't know or understand the volume lifecycle loops well enough to give feedback on this approach, it just doesn't seem principled to me, and seems like it could easily start trying to refresh volumes in pods we don't want it to.